pandas/tests/groupby/aggregate/test_cython.py

```diff
@@ -0,0 +1,180 @@
+                       'b': np.random.randint(0, 2, 50).astype('bool')})
+    result = frame.groupby('a')['b'].mean()
+    expected = frame.groupby('a')['b'].agg(np.mean)
+
+    tm.assert_series_equal(result, expected)
+
+
+def test_cython_agg_nothing_to_agg():
+    frame = DataFrame({'a': np.random.randint(0, 5, 50),
+                       'b': ['foo', 'bar'] * 25})
+    msg = "No numeric types to aggregate"
+
+    with tm.assert_raises_regex(DataError, msg):
+        frame.groupby('a')['b'].mean()
+
+    frame = DataFrame({'a': np.random.randint(0, 5, 50),
+                       'b': ['foo', 'bar'] * 25})
+    with tm.assert_raises_regex(DataError, msg):
+        frame[['b']].groupby(frame['a']).mean()
+
+
+def test_cython_agg_nothing_to_agg_with_dates():
+    frame = DataFrame({'a': np.random.randint(0, 5, 50),
+                       'b': ['foo', 'bar'] * 25,
+                       'dates': pd.date_range('now', periods=50, freq='T')})
+    msg = "No numeric types to aggregate"
+    with tm.assert_raises_regex(DataError, msg):
+        frame.groupby('b').dates.mean()
+
+
+def test_cython_agg_frame_columns():
+    # #2113
+    df = DataFrame({'x': [1, 2, 3], 'y': [3, 4, 5]})
+
+    df.groupby(level=0, axis='columns').mean()
+    df.groupby(level=0, axis='columns').mean()
+    df.groupby(level=0, axis='columns').mean()
+    df.groupby(level=0, axis='columns').mean()
+
+
+def test_cython_agg_return_dict():
+    # GH 16741
+    df = DataFrame(
+        {'A': ['foo', 'bar', 'foo', 'bar', 'foo', 'bar', 'foo', 'foo'],
+         'B': ['one', 'one', 'two', 'three', 'two', 'two', 'one', 'three'],
+         'C': np.random.randn(8),
+         'D': np.random.randn(8)})
+
+    ts = df.groupby('A')['B'].agg(lambda x: x.value_counts().to_dict())
+    expected = Series([{'two': 1, 'one': 1, 'three': 1},
+                       {'two': 2, 'one': 2, 'three': 1}],
+                      index=Index(['bar', 'foo'], name='A'),
+                      name='B')
+    tm.assert_series_equal(ts, expected)
+
+
+def test_cython_fail_agg():
+    dr = bdate_range('1/1/2000', periods=50)
+    ts = Series(['A', 'B', 'C', 'D', 'E'] * 10, index=dr)
+
+    grouped = ts.groupby(lambda x: x.month)
+    summed = grouped.sum()
+    expected = grouped.agg(np.sum)
+    tm.assert_series_equal(summed, expected)
+
+
+@pytest.mark.parametrize('op, targop', [
+    ('mean', np.mean),
+    ('median', np.median),
+    ('var', np.var),
+    ('add', np.sum),
+    ('prod', np.prod),
+    ('min', np.min),
+    ('max', np.max),
+    ('first', lambda x: x.iloc[0]),
+    ('last', lambda x: x.iloc[-1]),
+])
+def test__cython_agg_general(op, targop):
+    df = DataFrame(np.random.randn(1000))
+    labels = np.random.randint(0, 50, size=1000).astype(float)
+
+    result = df.groupby(labels)._cython_agg_general(op)
+    expected = df.groupby(labels).agg(targop)
+    try:
+        tm.assert_frame_equal(result, expected)
+    except BaseException as exc:
```

Comment:
yes u can take this try/except out
it’s helpful when not parametrized but not now